epigraphhub/analysis/clustering.py

```diff
@@ -0,0 +1,177 @@
+            cmat[i, j] = corr
+            lags[i, j] = lag
+    return cmat, lags
+
+
+# @st.cache
+def compute_clusters(
+    country,
+    curve,
+    columns,
+    t,
+    drop_georegions=None,
+    smooth=True,
+    ini_date=None,
+    plot=False,
+):
+    """
+    Function to compute the clusters od the data
+
+    params country: contry that we want to make the cluster.
+
+    param curve: string. Represent the curve that will used to cluster the regions.
+
+    param columns: list with 3 columns names in this order: ['date_colum',
+    'georegion_column', 'target_column']
+
+    param t: float. Represent the value used to compute the distance between the cluster and so
+    decide the number of clusters returned.
+
+    param drop_georegions: list. Param with the georegions that wiil be ignored in the
+    clusterization.
+
+    param smooth: Boolean. If true a rooling average of seven days will be applied to
+    the data.
+
+    param ini_date: Represent the initial date to start to compute the correlation
+    between the series.
+
+    param plot: Boolean. If true a dendogram of the clusterization will be returned.
+
+    return: array.
+
+    -> cluster: is the array with the computed clusters
+    -> all_regions: is the array with all the regions
+    """
+
+    country = country.lower()
+
+    df = get_georegion_data(country, "All", curve, columns)
+    df.index = pd.to_datetime(df[columns[0]])
+
+    inc_canton = df.pivot(columns=columns[1][1:-1], values=columns[2])
+
+    if smooth:
+        inc_canton = inc_canton.rolling(7).mean().dropna()
+
+    if ini_date:
+        inc_canton = inc_canton.loc[ini_date:]
+
+    if drop_georegions != None:
+
+        for i in drop_georegions:
+            del inc_canton[i]
+
+    # Computing the correlation matrix based on the maximum correlation lag
+    cm, lm = lag_ccf(inc_canton.values)
+
+    # Plotting the dendrogram
+    linkage = hcluster.linkage(cm, method="complete")
+
+    if plot:
+        fig, ax = plt.subplots(1, 1, figsize=(15, 10), dpi=300)
+        hcluster.dendrogram(
+            linkage, labels=inc_canton.columns, color_threshold=0.3, ax=ax
+        )
+        ax.set_title(
+            "Result of the hierarchical clustering of the series",
+            fontdict={"fontsize": 20},
+        )
+        plt.xticks(fontsize=16)
+        plt.yticks(fontsize=16)
+
+    else:
+        fig = None
+
+    # computing the cluster
+
+    ind = hcluster.fcluster(linkage, t, "distance")
+
+    grouped = pd.DataFrame(list(zip(ind, inc_canton.columns))).groupby(
+        0
+    )  # preciso entender melhor essa linha do código
+    clusters = [group[1][1].values for group in grouped]
+
+    all_regions = df.geoRegion.unique()
+
+    return clusters, all_regions, fig
+
+
+def plot_clusters(
+    country, curve, columns, clusters, ini_date=None, normalize=False, smooth=True
+):
+
+    country = country.lower()
+
+    df = get_georegion_data(country, "All", curve, columns)
```

Comment:
The plot function should not fetch data "again"! They are called by the function that have already loaded this data to calculate the clusters, therefore when they call plot_cluster the should pass that data along, avoiding another expensive query to the database.